test/unified_malloc_framework/memoryPool.hpp

```diff
@@ -84,6 +115,209 @@ TEST_P(umfPoolTest, pow2AlignedAlloc) {
+    ASSERT_NE(ptr, nullptr);
+
+    umfPoolFree(pool.get(), ptr);
+
+    for (auto allocation : allocations) {
+        umfPoolFree(pool.get(), allocation);
+    }
+}
+
+TEST_P(umfPoolTest, multiThreadedMallocFree) {
+    static constexpr size_t allocSize = 64;
+    auto poolMalloc = [](size_t allocSize, umf_memory_pool_handle_t pool) {
+        std::vector<void *> allocations;
+        for (size_t i = 0; i <= 10; ++i) {
+            allocations.emplace_back(umfPoolMalloc(pool, allocSize));
+            ASSERT_NE(allocations.back(), nullptr);
+        }
+
+        for (auto allocation : allocations) {
+            umfPoolFree(pool, allocation);
+        }
+    };
+
+    std::vector<std::thread> threads;
+    for (int i = 0; i < NTHREADS; i++) {
+        threads.push_back(std::thread(poolMalloc, allocSize, pool.get()));
+    }
+
+    for (auto &thread : threads) {
+        thread.join();
+    }
+}
+
+TEST_P(umfPoolTest, multiThreadedpow2AlignedAlloc) {
+#ifdef _WIN32
+    // TODO: implement support for windows
+    GTEST_SKIP();
+#endif
+
+    static constexpr size_t maxAlignment = (1u << 22);
+    static constexpr size_t numAllocs = 4;
+    auto poolPow2AlignedAlloc = [](size_t maxAlignment, size_t numAllocs,
+                                   umf_memory_pool_handle_t pool) {
+        for (size_t alignment = 1; alignment <= maxAlignment; alignment <<= 1) {
+            std::cout << alignment << std::endl;
+            std::vector<void *> allocs;
+
+            for (size_t alloc = 0; alloc < numAllocs; alloc++) {
+                auto *ptr = umfPoolAlignedMalloc(pool, alignment, alignment);
+                ASSERT_NE(ptr, nullptr);
+                ASSERT_TRUE(reinterpret_cast<uintptr_t>(ptr) % alignment == 0);
+                std::memset(ptr, 0, alignment);
+                allocs.push_back(ptr);
+            }
+
+            for (auto &ptr : allocs) {
+                umfPoolFree(pool, ptr);
+            }
+        }
+    };
+
+    std::vector<std::thread> threads;
+    for (int i = 0; i < NTHREADS; i++) {
+        threads.push_back(std::thread(poolPow2AlignedAlloc, maxAlignment,
+                                      numAllocs, pool.get()));
+    }
+
+    for (auto &thread : threads) {
+        thread.join();
+    }
+}
+
+TEST_P(umfPoolTest, multiThreadedReallocFree) {
+    static constexpr size_t allocSize = 64;
+    static constexpr size_t multiplier = 3;
+    auto poolRealloc = [](size_t allocSize, size_t multiplier,
+                          umf_memory_pool_handle_t pool) {
+        std::vector<void *> allocations;
+        for (size_t i = 0; i <= 10; ++i) {
+            allocations.emplace_back(umfPoolMalloc(pool, allocSize));
+            ASSERT_NE(allocations.back(), nullptr);
+        }
+
+        for (auto allocation : allocations) {
+            auto *ptr =
+                umfPoolRealloc(pool, allocation, allocSize * multiplier);
+            umf_result_t ret = umfPoolGetLastAllocationError(pool);
```

Comment:
I'm not sure if you should call GTEST_SKIP() from multiple threads. Perhaps it would be better to check the support once, and only then lunch the test? You could even implement some helpers for checking the support, something like this:

```
bool is_realloc_supported(umf_memory_pool_handle_t p) {
    static constexpr size_t allocSize = 1;
    bool supported;
    auto ptr = umfPoolMalloc(pool, allocSize);
    auto ptr2 = umfPoolRealloc(pool, ptr, allocSize * 2);
    
    if (ptr2) supported = true;
    else if (umfPoolGetLastAllocationError(pool) == UMF_RESULT_ERROR_NOT_SUPPORTED) supported = false;
    else throw std::runtime_error("realloc failed with unexpected error");
    
    umfPoolFree(ptr2);
}
```

Reply:
Done.